src/Microsoft.AspNet.Mvc.Razor/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -0,0 +1,206 @@
+        /// <param name="urlHelper">The <see cref="IUrlHelper"/>.</param>
+        /// <param name="htmlEncoder">The <see cref="IHtmlEncoder"/>.</param>
+        public UrlResolutionTagHelper(IUrlHelper urlHelper, IHtmlEncoder htmlEncoder)
+        {
+            UrlHelper = urlHelper;
+            HtmlEncoder = htmlEncoder;
+        }
+
+        /// <inheritdoc />
+        public override int Order
+        {
+            get
+            {
+                return int.MinValue;
+            }
+        }
+
+        protected IUrlHelper UrlHelper { get; }
+
+        protected IHtmlEncoder HtmlEncoder { get; }
+
+        /// <inheritdoc />
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            IEnumerable<string> attributeNames;
+            if (ElementAttributeLookups.TryGetValue(output.TagName, out attributeNames))
+            {
+                foreach (var attributeName in attributeNames)
+                {
+                    ProcessUrlAttribute(attributeName, output);
+                }
+            }
+
+            // itemid can be present on any HTML element.
+            ProcessUrlAttribute("itemid", output);
+        }
+
+        private void ProcessUrlAttribute(string attributeName, TagHelperOutput output)
+        {
+            IEnumerable<TagHelperAttribute> attributes;
+            if (output.Attributes.TryGetAttributes(attributeName, out attributes))
+            {
+                foreach (var attribute in attributes)
+                {
+                    string resolvedUrl;
+                    if (attribute.Value is string)
+                    {
+                        if (TryResolveUrl(
+                                (string)attribute.Value,
+                                tryEncodeApplicationPath: false,
+                                resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = resolvedUrl;
+                        }
+                    }
+                    else if (attribute.Value is HtmlString)
+                    {
+                        var htmlStringValue = ((HtmlString)attribute.Value).ToString();
+                        if (TryResolveUrl(
+                                htmlStringValue,
+                                tryEncodeApplicationPath: true,
+                                resolvedUrl: out resolvedUrl))
+                        {
+                            attribute.Value = new HtmlString(resolvedUrl);
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool TryResolveUrl(string url, bool tryEncodeApplicationPath, out string resolvedUrl)
+        {
+            resolvedUrl = null;
+
+            if (url == null)
+            {
+                return false;
+            }
+
+            var urlLength = url.Length;
+
+            // Find the start of the potential application relative URL value.
```

Comment:
@Eilon I must be confused but let's chat about this tomorrow.

What I'm seeing:
- Views (`.cshtml` files), precompiled files and a few other things are by default found using a `PhysicalFileProvider` rooted at `IApplicationEnvironment.ApplicationBasePath`.
 - Comments mentioning an "application relative path" are all about the above code.
- `FilePathResult`, the tag helpers for `<img>`, `<link>`, and `<script>`, and other classes dealing with "webroot" use `IHostingEnvironment.WebRootFileProvider` which is also an `IFileProvider` but rooted at the `IHostingEnvironment.WebRootPath`.
- At runtime in the MVC Sample, `IApplicationEnvironment.ApplicationBasePath` is `"C:\\...\\samples\\MvcSample.Web"` while `IHostingEnvironment.WebRootPath` is `"C:\\...\\samples\\MvcSample.Web\\wwwroot"`.